integrations/parquet/src/async_reader.rs

```diff
@@ -0,0 +1,343 @@
+        Self {
+            inner: reader,
+            content_len,
+            footer_size: set_footer_size(PREFETCH_FOOTER_SIZE, content_len),
+        }
+    }
+
+    /// Set prefetch size for fetching file footer.
+    pub fn with_footer_size(mut self, footer_size: usize) -> Self {
+        self.footer_size = set_footer_size(footer_size, self.content_len);
+        self
+    }
+}
+
+impl AsyncFileReader for AsyncReader {
+    fn get_bytes(&mut self, range: Range<usize>) -> BoxFuture<'_, ParquetResult<bytes::Bytes>> {
+        async move {
+            Ok(self
+                .inner
+                .read(range.start as u64..range.end as u64)
+                .await
+                .map_err(|err| ParquetError::External(Box::new(err)))?
+                .to_bytes())
+        }
+        .boxed()
+    }
+
+    fn get_byte_ranges(
+        &mut self,
+        ranges: Vec<Range<usize>>,
+    ) -> BoxFuture<'_, ParquetResult<Vec<bytes::Bytes>>> {
+        async move {
+            Ok(self
+                .inner
+                .fetch(
+                    ranges
+                        .into_iter()
+                        .map(|range| range.start as u64..range.end as u64)
+                        .collect(),
+                )
+                .await
+                .map_err(|err| ParquetError::External(Box::new(err)))?
+                .into_iter()
+                .map(|buf| buf.to_bytes())
+                .collect::<Vec<_>>())
+        }
+        .boxed()
+    }
+
+    fn get_metadata(&mut self) -> BoxFuture<'_, ParquetResult<std::sync::Arc<ParquetMetaData>>> {
+        async move {
+            let prefetched_footer = self
+                .inner
+                .read(self.content_len - self.footer_size as u64..self.content_len)
+                .await
+                .map_err(|err| ParquetError::External(Box::new(err)))?;
+            let prefetched_footer_len = prefetched_footer.len();
+            let footer_len =
+                &prefetched_footer.slice((prefetched_footer_len - 8)..prefetched_footer_len);
+
+            // Safety: checked above.
+            let buf: [u8; 8] = footer_len.chunk().try_into().unwrap();
+            let metadata_len = decode_footer(&buf)?;
+
+            let buf = if prefetched_footer_len >= metadata_len + FOOTER_SIZE {
+                prefetched_footer.slice(
+                    (prefetched_footer_len - metadata_len - FOOTER_SIZE)
+                        ..(prefetched_footer_len - FOOTER_SIZE),
+                )
+            } else {
+                self.inner
+                    .read(
+                        self.content_len - metadata_len as u64 - FOOTER_SIZE as u64
+                            ..self.content_len - FOOTER_SIZE as u64,
+                    )
+                    .await
+                    .map_err(|err| ParquetError::External(Box::new(err)))?
+            };
+
+            Ok(Arc::new(decode_metadata(buf.chunk())?))
```

Comment:
Hi, please note that `buf.chunk()` only returns the first chunk in `Buffer`. We might need to use `to_vec` to obtain the full content.

Reply:
Updated.